src/examples/huggingface/convert_checkpoint_from_hf.py

```diff
@@ -8,165 +8,372 @@
+    transformer_config_dict: Dict[str, Any],
+    tokenizer_config_dict: Dict[str, Any],
+    *,
+    max_sequence_length: int = -1,
+    validate: bool = True,
+    debug: bool = False,
+    device: torch.device | None = None,
+) -> None:
+    """
+    Convert a HF checkpoint to an OLMo core checkpoint.
+
+    Args:
+        hf_checkpoint_path: Path to the original HF checkpoint
+        output_path: Where to save the converted model
+        transformer_config_dict: Dictionary form of OLMo core model config
+        tokenizer_config_dict: Dictionary form of OLMo core tokenizer config
+    """
+    if max_sequence_length <= 0:
+        raise ValueError(f"Missing or invalid sequence length: {max_sequence_length}")
+
+    # Remove deprecated transformer config options
+    if "compile" in transformer_config_dict:
+        del transformer_config_dict["compile"]
+    if "dp_config" in transformer_config_dict:
+        del transformer_config_dict["dp_config"]
+    if "tp_config" in transformer_config_dict:
+        del transformer_config_dict["tp_config"]
+    if "float8_config" in transformer_config_dict:
+        del transformer_config_dict["float8_config"]
+
+    model = TransformerConfig.from_dict(transformer_config_dict).build()
+
+    # Replace weight init with an efficient alternative that just allocates memory
+    @torch.no_grad()
+    def init_weights(
+        self: Transformer,
+        *,
+        max_seq_len: Optional[int] = None,
+        max_local_microbatch_size: Optional[int] = None,
+        device: Optional[torch.device] = None,
+        pp_mesh: Optional[DeviceMesh] = None,
+    ) -> torch.Generator:
+        """
+        Initialize the model weights.
+
+        :param max_seq_len: The maximum sequence length expected. This is used
+            to warm up the RoPE cache.
+        :param max_local_microbatch_size: The maximum local (rank) micro-batch size (in tokens)
+            expected. This is used to warm-up some MoE cache.
+        :param device: The device the local copy of the model will be trained on.
+        :param pp_mesh: Pipeline parallel mesh. Pass this when using pipeline parallelism
+            to ensure the weights are initialized differently for different stages.
+        """
+        device = device or self.device
+        self.to_empty(device=device)
+
+        for module in self.modules():
+            if hasattr(module, "reset_parameters"):
+                module.to_empty(device=device)  # type: ignore
+
+        seed = self.init_seed
+        if pp_mesh is not None:
+            seed += pp_mesh.get_local_rank()
+        return torch.Generator(device).manual_seed(seed)
+
+    model.init_weights = types.MethodType(init_weights, model)
+
+    device = device or get_default_device()
+
+    train_module = TransformerTrainModuleConfig(
+        rank_microbatch_size=max_sequence_length,
+        max_sequence_length=max_sequence_length,
+        optim=AdamWConfig(),
+    ).build(model, device=device)
```

Comment:
I know technically the train module decides the format of the checkpoint, but it seems overkill to have to instantiate a train module here. In this case I think it's fair to make assumptions about the format and just use the corresponding functions from `olmo_core.distributed.checkpoint` directly.